lib/fastly_nsq/cli.rb

```diff
@@ -0,0 +1,238 @@
+    pid = ::File.read(pidfile).to_i
+    return :dead if pid == 0
+    Process.kill(0, pid) # check process status
+    :running
+  rescue Errno::ESRCH
+    :dead
+  rescue Errno::EPERM
+    :not_owned
+  end
+
+  def setup_logger
+    FastlyNsq.logger = Logger.new(options[:logfile]) if options[:logfile]
+
+    FastlyNsq.logger.level = ::Logger::DEBUG if options[:verbose]
+  end
+
+  def write_pid
+    if pidfile?
+      begin
+        File.open(pidfile, ::File::CREAT | ::File::EXCL | ::File::WRONLY) do |f|
+          f.write Process.pid.to_s
+        end
+        at_exit { File.delete(pidfile) if File.exist?(pidfile) }
+      rescue Errno::EEXIST
+        check_pid
+        retry
+      end
+    end
+  end
+
+  def trap_signals
+    self_read, self_write = IO.pipe
+    sigs = %w(INT TERM TTIN USR1)
+
+    sigs.each do |sig|
+      begin
+        trap sig do
+          self_write.puts(sig)
+        end
+      rescue ArgumentError
+        puts "Signal #{sig} not supported"
+      end
+    end
+    self_read
+  end
+
+  def handle_signal(sig)
+    FastlyNsq.logger.debug "Got #{sig} signal"
+    case sig
+    when 'INT'
+      # Handle Ctrl-C in JRuby like MRI
+      # http://jira.codehaus.org/browse/JRUBY-4637
+      raise Interrupt
+    when 'TERM'
+      # Heroku sends TERM and then waits 10 seconds for process to exit.
+      raise Interrupt
+    when 'USR1'
+      FastlyNsq.logger.info 'Received USR1, no longer accepting new work'
+      launcher.quiet
+    when 'TTIN'
+      handle_ttin
+    end
+  end
+
+  def handle_ttin
+    Thread.list.each do |thread|
+      FastlyNsq.logger.warn "Thread TID-#{thread.object_id.to_s(36)} #{thread['fastly_nsq_label']}"
+      if thread.backtrace
+        FastlyNsq.logger.warn thread.backtrace.join("\n")
+      else
+        FastlyNsq.logger.warn '<no backtrace available>'
+      end
+    end
+  end
+
+  def damonize
+    return unless options[:daemonize]
+
+    files_to_reopen = []
+    ObjectSpace.each_object(File) do |file|
+      files_to_reopen << file unless file.closed?
+    end
+
+    ::Process.daemon(true, true)
+
+    reopen(files_to_reopen)
+
+    [$stdout, $stderr].each do |io|
+      File.open(options.fetch(:logfile, '/dev/null'), 'ab') do |f|
+        io.reopen(f)
+      end
+      io.sync = true
+    end
+    $stdin.reopen('/dev/null')
+
+    setup_logger
+  end
+
+  def reopen(files)
+    files.each do |file|
+      begin
+        file.reopen file.path, 'a+'
+        file.sync = true
+      rescue IOError => e
+        FastlyNsq.logger.warn "IOError reopening file:  #{e.message}"
+      rescue StandardError => e
+        FastlyNsq.logger.error "Non IOError reopening file:  #{e.message}"
+      end
+    end
+  end
+
+  def logfile?
+    !logfile.nil?
+  end
+
+  def pidfile?
+    !pidfile.nil?
+  end
+
+  def damonize?
+    options[:daemonize]
+  end
+
+  def logfile
+    options[:logfile]
+  end
+
+  def pidfile
+    options[:pidfile]
+  end
+end
```

Comment:
The cli is really a daemonized listener / multi-threaded consumer. Maybe the name should reflect that?